### ISSUE TITLE
[Parity 13/16] Nested / Named / Ragged / Masked tensors

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#222]
Part of epic #209.

## Summary

PyTorch has three special tensor subtypes that handle structured shape variability: **NestedTensor** (variable-length sequences without padding), **Named Tensor** (axis labels for safer indexing and broadcasting), and **MaskedTensor** (tensors with explicit masked-out positions). We have none of them. These are especially valuable for NLP, RecSys, and graph workloads where padding + mask bookkeeping otherwise dominates user code.

### NestedTensor (`torch.nested`)
- `torch.nested.nested_tensor(list_of_tensors, dtype=, device=, layout=jagged|strided)` — variable-length sequences
- Ops: `linear`, `matmul`, `add`, `softmax`, `log_softmax`, `layer_norm`, `mha`, `scaled_dot_product_attention`
- Lazy padding — materializes only if requested
- `.values()`, `.offsets()`, `.to_padded_tensor(padding, output_size)`
- Jagged layout (single underlying storage + offsets) + strided layout

### Named Tensor
- `torch.ones(32, 256, names=("batch", "features"))`
- Name-aware ops: `.align_to("batch", "features")`, `.rename(...)`, `.refine_names(...)`
- Name-aware broadcasting
- `.flatten(["height", "width"], "hw")` — name-collapse
- `.names` attribute, `.has_names()`

### MaskedTensor (`torch.masked`)
- `masked_tensor(data, mask, requires_grad)`
- All elementwise ops ignore masked-out positions
- `sum`, `mean`, `std`, `var`, `argmin`, `argmax` respect mask
- Reductions return masked results when all inputs in a slice are masked
- Addresses the long-standing `nan` / `inf` reduction correctness issue

## Current state in the repo

- Nothing. `Tensor<T>` is dense-only.
- `TensorShape.cs` and `TensorLayoutHint.cs` handle dense layouts.

## How we beat PyTorch

1. **Jagged NestedTensor as a first-class `TensorView<T>`**: we already have views. A jagged view is storage + offsets + per-row lengths — zero new storage class. PyTorch's NestedTensor is a separate subclass with partial op coverage (the "supported ops" list has holes).
2. **All attention ops on jagged inputs without padding**: FlashAttention-2 (already shipped) + jagged offsets + block-sparse mask = native packed-sequence attention. PyTorch supports `F.scaled_dot_product_attention` on NestedTensor but only via one kernel; we ship all three (math / flash / mem-efficient) for nested.
3. **Named tensor names that survive compile**: `LazyTensorScope` nodes carry name metadata through fusion. PyTorch's named tensors are erased on `torch.compile` boundary.
4. **MaskedTensor without double-memory**: our mask is a bitpacked view that shares storage with the mask tensor. PyTorch stores a full bool tensor.
5. **Masked-aware ops through `OpRegistry`**: an op registered with a masked variant is automatically picked when input is `MaskedTensor<T>`.
6. **Performance**: for tasks like variable-length batched attention, the nested path avoids padding FLOPs entirely. Target 30–50% throughput win on a batched-variable-length benchmark vs PyTorch's NestedTensor path.
7. **Round-trip bridges**: `NestedTensor ↔ padded tensor + mask ↔ masked tensor` with zero copy (all share underlying storage).

### NestedTensor
- [ ] `NestedTensor<T>` as a `TensorView<T>` variant with offsets + per-row lengths
- [ ] Jagged layout + strided layout
- [ ] `NestedTensor.FromList(tensors)` constructor
- [ ] `.Values()`, `.Offsets()`, `.ToPadded(padding, outputSize)`
- [ ] Ops: `Linear`, `MatMul`, `Add`, `Softmax`, `LogSoftmax`, `LayerNorm`, `GeLU`, `SiLU`, `ScaledDotProductAttention` (via existing FlashAttention-2), `MultiHeadAttention`
- [ ] Backward for all above
- [ ] Zero-copy bridge to `MaskedTensor<T>`

### Named Tensor
- [ ] `NamedDim<T>` annotation on `Tensor<T>` (names array + has_names flag)
- [ ] `.Rename(...)`, `.RefineNames(...)`, `.AlignTo(...)`, `.AlignAs(other)`
- [ ] Name-aware broadcasting (dim matching by name, not position)
- [ ] `.Flatten(names[], newName)`, `.Unflatten(name, sizes)`
- [ ] Names survive graph compile (`LazyTensorScope`)

### MaskedTensor
- [ ] `MaskedTensor<T>` with bitpacked mask storage
- [ ] Elementwise ops ignore masked-out positions
- [ ] Reductions: `Sum`, `Mean`, `Std`, `Var`, `ArgMin`, `ArgMax`, `Min`, `Max`, `Prod` — all mask-aware
- [ ] Backward routes through mask
- [ ] `OpRegistry` dispatch for MaskedTensor inputs

## Acceptance criteria

- NestedTensor benchmark: variable-length batched attention ≥30% faster than PyTorch NestedTensor + SDPA at realistic length-distribution.
- Named tensor broadcast: test that mismatched positional but matching names works; test that dimension-name contradictions throw.
- MaskedTensor reduction: `mean` over a tensor with some masked positions matches a manual boolean-indexed reference.
- Round-trip bridges: `padded → nested → padded` is equality-preserving; `masked → dense` with default fill is equality-preserving.
- Autograd gradcheck passes for every op on all three types.

## Implementation notes

- New files under `src/AiDotNet.Tensors/LinearAlgebra/Nested/`, `LinearAlgebra/Named/`, `LinearAlgebra/Masked/`.
- Jagged NestedTensor reuses `TensorView<T>` + `TensorStorage<T>` — add an offsets field.
- Named metadata goes on `TensorShape.cs` (optional `Names` property) and threaded through `LazyNode`.
- MaskedTensor bitpacked mask reuses `NumericOperations/PackedInt1.cs` representation.
- OpRegistry extension point: dispatch by tensor subtype, not just dtype.

**Comment by ooples** [comment]
# Implementation plan — detailed

## 1. Current state

Nothing. `Tensor<T>` is dense-only. `TensorShape.cs` + `TensorLayoutHint.cs` handle dense layouts.

## 2. Scope — full inventory

### NestedTensor
```
NestedTensor<T> as TensorView<T> variant with offsets + per-row lengths
Jagged layout (single storage + offsets) + strided layout
NestedTensor.FromList(tensors)
.Values(), .Offsets(), .ToPadded(padding, outputSize)
Ops: Linear, MatMul, Add, Softmax, LogSoftmax, LayerNorm, GeLU, SiLU,
     ScaledDotProductAttention (via existing FlashAttention-2), MultiHeadAttention
Backward for all above
Zero-copy bridge to MaskedTensor<T>
```

### Named Tensor
```
NamedDim<T> annotation on Tensor<T> (names array + has_names flag)
.Rename(...), .RefineNames(...), .AlignTo(...), .AlignAs(other)
Name-aware broadcasting (match by name, not position)
.Flatten(names[], newName), .Unflatten(name, sizes)
Names survive graph compile (LazyTensorScope)
```

### MaskedTensor
```
MaskedTensor<T> with bitpacked mask storage
Elementwise ops ignore masked-out positions
Reductions: Sum, Mean, Std, Var, ArgMin, ArgMax, Min, Max, Prod — mask-aware
Backward routes through mask
OpRegistry dispatch for MaskedTensor inputs
```

**Total: ~35 surfaces across three tensor types.**

## 3. Win vectors

1. **Jagged NestedTensor as first-class `TensorView<T>`** — storage + offsets + lengths; zero new storage class. PyTorch's NestedTensor is a separate subclass with gaps.
2. **All attention ops on jagged without padding**: FlashAttention-2 + jagged offsets + block-sparse mask = native packed-sequence attention. We ship math/flash/mem-efficient variants.
3. **Named tensor names survive compile**: `LazyTensorScope` nodes carry name metadata through fusion. PyTorch erases names at `torch.compile` boundary.
4. **MaskedTensor without double-memory**: bitpacked view sharing storage. PyTorch stores full bool.
5. **Masked-aware ops through `OpRegistry`** — auto-dispatch on `MaskedTensor<T>` input.
6. **Performance on variable-length**: jagged avoids padding FLOPs. Target 30–50% vs PyTorch NestedTensor + SDPA.
7. **Round-trip bridges** NestedTensor ↔ padded+mask ↔ MaskedTensor zero-copy.

## 4. Per-axis beat targets

| Axis | Target |
|---|---|
| Variable-length batched attention NestedTensor | ≥30% vs PyTorch NestedTensor + SDPA |
| Named broadcast correctness | mismatched pos but matching names → works |
| MaskedTensor reduction correctness | mask-aware `mean` matches bool-indexed ref |
| Round-trip padded↔nested | equality-preserving |
| MaskedTensor memory overhead | ~1/8× of bool (bitpacked) |
| Autograd | gradcheck for every op on all three types |
| Compile | names survive LazyTensorScope fusion |
| All 6 GPU backends | CUDA required; others best-effort |

## 5. File inventory

**New (~50):**
- `src/AiDotNet.Tensors/LinearAlgebra/Nested/` — NestedTensor, Jagged, Strided, ToPadded, FromList, NestedOps (~12)
- `src/AiDotNet.Tensors/LinearAlgebra/Named/` — NamedDim, NamedBroadcast, Rename, RefineNames, AlignTo, Flatten, Unflatten (~10)
- `src/AiDotNet.Tensors/LinearAlgebra/Masked/` — MaskedTensor, BitpackedMask, MaskedReductions, MaskedElementwise (~10)
- Masked/nested-aware op variants (~15)
- GPU kernels for nested (~10)

**Modified (~12):** `TensorShape.cs` (Names), `TensorView.cs`, `TensorStorage.cs` (offsets field), `OpRegistry.cs` (subtype dispatch), `LazyTensorScope.cs`, `LazyNode.cs`, `DifferentiableOps.cs`, `FlashAttention.cs` (jagged input support), `BackwardFunctions.cs`, `PackedInt1.cs` (bitpacked mask reuse).

**Tests (~25):** NestedTensor op coverage + gradcheck, named broadcast correctness, MaskedTensor reduction correctness, round-trip bridges, jagged-attention parity vs padded.

**Benchmarks (~4):** `NestedAttentionBenchmarks`, `NamedTensorCompileBenchmarks`, `MaskedReductionBenchmarks`, `RoundTripBridgeBenchmarks`.

## 6. Acceptance gate

1. NestedTensor op coverage (Linear/MatMul/Add/Softmax/LayerNorm/SDPA/MHA) with gradcheck.
2. Variable-length batched attention ≥30% faster than PyTorch NestedTensor+SDPA.
3. Named broadcast: mismatched positional but matching names works; contradictions throw.
4. MaskedTensor reduction correctness vs bool-indexed ref.
5. Round-trip bridges equality-preserving.
6. Named metadata survives LazyTensorScope fusion.
7. All three types differentiable.

## 7. Timeline — ~5 weeks

W1 NestedTensor jagged + FromList + Values/Offsets/ToPadded + basic ops (Add/Linear/MatMul).
W2 NestedTensor softmax + layernorm + attention (reuse FlashAttention-2 + jagged offsets).
W3 Named Tensor + broadcast + compile-preservation.
W4 MaskedTensor + bitpacked mask + mask-aware reductions + elementwise.
W5 Round-trip bridges + benchmarks + PR.

## 8. Dependencies

- **Blocks**: any future NLP demos (transformers without padding).
- **Blocked by**: #210 (movement ops for offsets manipulation), #214 (autograd for new subtype dispatch).

## 9. Risks

- **OpRegistry subtype dispatch** — interacts with every op; audit burden.
- **Jagged FlashAttention integration** — offsets handling in kernel; testing on realistic variable-length distributions.
- **Named broadcasting semantics** — PyTorch's rules are subtle; pin to a named-style spec document.

---

*Part of epic #209. Plan ratified 2026-04-19.*
